honeybee_energy/material/_base.py

```diff
@@ -0,0 +1,91 @@
+# coding=utf-8
+"""Base energy material."""
+from __future__ import division
+
+from honeybee.typing import valid_ep_string
+
+
+class _EnergyMaterialBase(object):
+    """Base energy material.
+
+    Properties:
+        name
+    """
+    _name_descr = 'material name'
+    _is_window_material = None
```

Comment:
should we set this in instance level instead of class level? If one is not careful this can end up being an issue. What is the advantage of having it as a class attribute?

Reply:
I see similar design pattern in other classes. let's discuss it in one place here.

Reply:
It is a hidden attribute that is exposed as a read-only property.  The reason why its not set on the instance is that it needs to be fundamentally tied to the class (it is basically an indicator of what constructions the material class can be a part of).  The main reason why it is there is because I hid the base classes for different types of materials (the base classes do not have enough to be simulate-able and it didn't feel right to expose them). So this property allows users to check the type of material without needing to access the hidden base classes to make `isinstance` calls.

You can see that I don't use this property anywhere in the rest of the API because I know where the hidden classes are and I just use `isinstance`.  However, I feel that it is important to include for the users of the honeybee energy library.

Reply:
For some reason GitHub doesn't show me the full file. I can follow your logic but you can keep it in instance and then expose it as property and it will do the same. As you said in the near future more people should be able to understand the logic and this is one of those cases that being explicit will be probably better.

```python
class _base(object):
    def __init__(self):
        self._is_window_material = False

class WindowMaterial(_base):
    def __init__(self):
         _base.__init__(self)
        # now set to False
        self._is_window_material = False

class OpaqueMaterial(_base):
    def __init__(self):
         _base.__init__(self)
        # It is good as is.
```

Reply:
> you can keep it in instance and then expose it as property and it will do the same
... yes except that I will have to add it to `__slots__` and it will take up a bit more memory since every instance will have its own version of this property.  But that's not as big of a reason as the fact that is property is supposed to be a marker of the inheritance tree of objects under the hood. I see that your example is trying to preserve this by always calling `__init__` of the parent class in the child class but it is not clear to developers that, by calling `_base._init()` they need to add the `_is_window_material` property to `__slots__`.
And I still don't understand what the issue is with the code that is there now.  Are you worried about someone overwriting the hidden property that is stored on the parent class?  If so, why is it that you are worried about people overwriting this hidden property in particular and not any of the other hidden properties like `__slots__`?  I think people generally recognize that hidden properties should not be overwritten unless they know exactly what they are doing.

Reply:
>  I see that your example is trying to preserve this by always calling __init__ of the parent class in the child class but it is not clear to developers that, by calling _base._init() they need to add the _is_window_material property to `__slots__`.

They don't have to add it to `__slots__` unless they want to use `__slots__` for the subclass which in that case this is not any different from other attributes. Here is the updated example.

```python
class _base(object):
    __slots__ = ('_is_window_material',)
    def __init__(self):
        self._is_window_material = False

class WindowMaterial(_base):
    __slots__ = ('_is_window_material',)
    def __init__(self):
         _base.__init__(self)
        # now set to False
        self._is_window_material = False

class OpaqueMaterial(_base):
    __slots__ = ()    
    def __init__(self):
         _base.__init__(self)
        # It is good as is.
```

Reply:
As discussed, I went with the same setup that you have in `honeybee_radiance`